File: BridgeX/app/src/main/java/cz/lriedel/bridgex/PermissionManager.java
package cz.lriedel.bridgex;

import android.content.pm.PackageManager;
import android.os.Build;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class PermissionManager {

    private static final int REQUEST_CODE_NOTIFICATIONS = 1001;
    private static final int REQUEST_CODE_FOREGROUND_LOCATION = 1002;
    private static final int REQUEST_CODE_BACKGROUND_LOCATION = 1003;


    private final MainActivity mainActivity;

    public PermissionManager(MainActivity mainActivity) {
        this.mainActivity = mainActivity;
    }

    public void requestAllPermissions(boolean silentMode) {
        if (!requestNotificationPermission()) {
            return;
        }
        if (!requestForegroundLocationPermission()) {
            return;
        }
        if (!requestBackgroundLocationPermission()) {
            return;
        }

        if (!silentMode) {
            Toast.makeText(mainActivity, R.string.all_permissions_granted, Toast.LENGTH_SHORT).show();
        }
    }

    private boolean requestNotificationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(mainActivity, android.Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(mainActivity,
                        new String[]{ android.Manifest.permission.POST_NOTIFICATIONS }, REQUEST_CODE_NOTIFICATIONS);
                return false;
            }
        }
        return true;
    }

    private boolean requestForegroundLocationPermission() {
        if (ContextCompat.checkSelfPermission(mainActivity, android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                || ContextCompat.checkSelfPermission(mainActivity, android.Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(mainActivity,
                    new String[]{ android.Manifest.permission.ACCESS_FINE_LOCATION, android.Manifest.permission.ACCESS_COARSE_LOCATION }, REQUEST_CODE_FOREGROUND_LOCATION);
            return false;
        }
        return true;
    }

    private boolean requestBackgroundLocationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            if (ContextCompat.checkSelfPermission(mainActivity, android.Manifest.permission.ACCESS_BACKGROUND_LOCATION)
                    != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(mainActivity,
                        new String[]{ android.Manifest.permission.ACCESS_BACKGROUND_LOCATION }, REQUEST_CODE_BACKGROUND_LOCATION);
                return false;
            }
        }
        return true;
    }

    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == REQUEST_CODE_NOTIFICATIONS) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(mainActivity, R.string.notifications_granted, Toast.LENGTH_SHORT).show();
                requestAllPermissions(false);
            }
            else {
                Toast.makeText(mainActivity, R.string.notifications_denied, Toast.LENGTH_SHORT).show();
            }
        }
        else if (requestCode == REQUEST_CODE_FOREGROUND_LOCATION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(mainActivity, R.string.foreground_location_granted, Toast.LENGTH_SHORT).show();
                requestAllPermissions(false);
            }
            else {
                Toast.makeText(mainActivity, R.string.foreground_location_denied, Toast.LENGTH_SHORT).show();
            }
        }
        else if (requestCode == REQUEST_CODE_BACKGROUND_LOCATION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(mainActivity, R.string.background_location_granted, Toast.LENGTH_SHORT).show();
                requestAllPermissions(false);
            }
            else {
                Toast.makeText(mainActivity, R.string.background_location_denied, Toast.LENGTH_SHORT).show();
            }
        }
    }
}

File: BridgeX/app/src/main/java/cz/lriedel/bridgex/PermissionManager.kt
package cz.lriedel.bridgex

import android.Manifest
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.health.connect.client.HealthConnectClient
import androidx.health.connect.client.PermissionController
import androidx.health.connect.client.permission.HealthPermission
import androidx.health.connect.client.records.DistanceRecord
import androidx.health.connect.client.records.StepsRecord
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.launch

class PermissionManager(
    private val mainActivity: MainActivity
) {
    private val healthClient = HealthConnectClient.getOrCreate(mainActivity)
    private val requiredHealthPermissions = arrayOf(
        HealthPermission.getReadPermission(StepsRecord::class.java.kotlin),
        HealthPermission.getReadPermission(DistanceRecord::class.java.kotlin)
    )
    private val healthPermissionLauncher = mainActivity.registerForActivityResult(
        PermissionController.createRequestPermissionResultContract()) {}

    // TODO: Make sure permissions are requested only once.
    // TODO: Make sure the application can operate even when missing some of the permissions.
    fun requestAllPermissions() {
        if (!requestNotificationPermission()) {
            return
        }
        if (!requestForegroundLocationPermission()) {
            return
        }
        if (!requestBackgroundLocationPermission()) {
            return
        }
        if (!requestHealthPermissions()) {
            return
        }
    }

    private fun requestNotificationPermission(): Boolean {
        if (ContextCompat.checkSelfPermission(mainActivity, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(mainActivity, arrayOf(Manifest.permission.POST_NOTIFICATIONS), REQUEST_CODE_NOTIFICATIONS)
            return false
        }
        return true
    }

    private fun requestForegroundLocationPermission(): Boolean {
        if (ContextCompat.checkSelfPermission(mainActivity, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
            || ContextCompat.checkSelfPermission(mainActivity, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(mainActivity, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION), REQUEST_CODE_FOREGROUND_LOCATION
            )
            return false
        }
        return true
    }

    private fun requestBackgroundLocationPermission(): Boolean {
        if (ContextCompat.checkSelfPermission(mainActivity, Manifest.permission.ACCESS_BACKGROUND_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(mainActivity, arrayOf(Manifest.permission.ACCESS_BACKGROUND_LOCATION),
                REQUEST_CODE_BACKGROUND_LOCATION)
            return false
        }
        return true
    }

    private fun requestHealthPermissions(): Boolean {
        mainActivity.lifecycleScope.launch {
            val granted = healthClient.permissionController.getGrantedPermissions()
            if (!granted.containsAll(requiredHealthPermissions.toList())) {
                healthPermissionLauncher.launch(requiredHealthPermissions.toSet())
            }
        }

        if (ContextCompat.checkSelfPermission(mainActivity, Manifest.permission.ACTIVITY_RECOGNITION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(mainActivity, arrayOf(Manifest.permission.ACTIVITY_RECOGNITION),
                REQUEST_CODE_ACTIVITY_RECOGNITION)
            return false
        }
        return true
    }

    fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == REQUEST_CODE_NOTIFICATIONS) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                requestAllPermissions()
            }
        }
        else if (requestCode == REQUEST_CODE_FOREGROUND_LOCATION) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                requestAllPermissions()
            }
        }
        else if (requestCode == REQUEST_CODE_BACKGROUND_LOCATION) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                requestAllPermissions()
            }
        }
    }

    companion object {
        private const val REQUEST_CODE_NOTIFICATIONS = 1001
        private const val REQUEST_CODE_FOREGROUND_LOCATION = 1002
        private const val REQUEST_CODE_BACKGROUND_LOCATION = 1003
        private const val REQUEST_CODE_ACTIVITY_RECOGNITION = 1004
    }
}

